### PR TITLE
Yarn - Fix incorrect impact graph

### PR DESCRIPTION
**Pull request by Or-Geva**
- [ ] All [tests](https://github.com/jfrog/jfrog-vscode-extension#building-and-testing-the-sources) passed. If this feature is not already covered by the tests, I added new tests.
- [ ] I used `npm run format` for formatting the code before submitting the pull request.
-----
Prior to this PR, there was an issue with the way Yarn calculated the impact graph. The problem intruduced from using the 'yarn list' output. The 'yarn list' output did not accurately represent the project's dependencies as it removed duplicate leaves and flattened the tree structure.

In order to resolve this issue, this PR changes the following:

1. Instead of constructing the entire dependency tree, we now compute only a partial tree. This partial tree specifically represents the paths from the project's root to the vulnerable dependencies only. This modification eliminates the need to calculate the entire tree, which not only reduces memory consumption but also aligns with our goal of displaying **only** paths to vulnerable dependencies.

2. Rather than relying on `yarn list`, we have switched to using `yarn why` which provides the correct path to a specific dependency. This involves gathering all vulnerable dependencies and utilizing 'yarn why' to map out the path of each dependency from the root.

### CODE REVIEW

File: src/main/treeDataProviders/utils/yarnImpactGraph.ts [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+
+        // Merge the children recursively
+        if (root1.children && root2.children) {
+            const mergedChildren: IImpactGraphNode[] = [];
+
+            for (const child1 of root1.children) {
+                const matchingChild2: IImpactGraphNode | undefined = root2.children.find(child2 => child2.name === child1.name);
+
+                if (matchingChild2) {
+                    const tmp: IImpactGraphNode | null = this.mergeTrees(child1, matchingChild2);
```
Comment (yahavi) [inline src/main/treeDataProviders/utils/yarnImpactGraph.ts:188]: Unmeaningful name "tmp"

File: src/main/treeDataProviders/utils/yarnImpactGraph.ts [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+                    if (tmp) {
+                        mergedChildren.push(tmp);
+                    }
+                } else {
+                    mergedChildren.push(child1); // If not found in root2, keep the child from root1
```
Comment (yahavi) [inline src/main/treeDataProviders/utils/yarnImpactGraph.ts:193]: Our convention is to put comments above the line

File: src/main/treeDataProviders/utils/yarnImpactGraph.ts [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+                }
+            }
+
+            for (const child2 of root2.children) {
+                const matchingChild1: IImpactGraphNode | undefined = root1.children.find(child1 => child1.name === child2.name);
+
+                if (!matchingChild1) {
+                    mergedChildren.push(child2); // Add children from root2 that are not present in root1
```
Comment (yahavi) [inline src/main/treeDataProviders/utils/yarnImpactGraph.ts:201]: Our convention is to put comments above the line

File: src/main/treeDataProviders/utils/yarnImpactGraph.ts [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+    public mergeTrees(root1: IImpactGraphNode | null, root2: IImpactGraphNode | null): IImpactGraphNode | null {
+        if (!root1 && !root2) {
+            return null;
+        }
+        if (!root1) {
+            return root2;
+        }
+        if (!root2) {
+            return root1;
+        }
```
Comment (yahavi) [inline src/main/treeDataProviders/utils/yarnImpactGraph.ts:175]: Shorter
```suggestion
        if (!root1 || !root2) {
            return root1 || root2;
        }
```

File: src/test/tests/yarnImpactGraph.test.ts [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+import { assert } from 'chai';
+import { RootNode } from '../../main/treeDataProviders/dependenciesTree/dependenciesRoot/rootTree';
+
+describe('Yarn impact graph util', async () => {
+    it('Build single impact graph', async () => {
+        const results: IImpactGraph = new YarnImpactGraphUtilMock('minimist', '0.0.8', 'Mock-Project', '').create();
+        assert.deepEqual(results, generateExpectedSingleImpactGraph());
+    });
+
+    it.only('Build multiple impact graphs', async () => {
```
Comment (yahavi) [inline src/test/tests/yarnImpactGraph.test.ts:12]: Is the "only" really needed?

**Review by yahavi — CHANGES_REQUESTED** [review]

**Review by yahavi — CHANGES_REQUESTED** [review]

File: src/main/treeDataProviders/utils/yarnImpactGraph.ts [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+     */
+    constructor(private dependencyName: string, private dependencyVersion: string, private projectName: string, private workspaceFolder: string) {}
+
+    /**
+     * Creates and returns an impact graph based on "yarn why" command output.
+     * @returns An impact graph.
+     */
+    public create(): IImpactGraph {
+        const dependencyChain: string[] | undefined = this.findDependencyChain(this.runYarnWhy());
+        if (dependencyChain) {
```
[the remark this answers is not in the document]
Comment (Or-Geva) [reply src/main/treeDataProviders/utils/yarnImpactGraph.ts:49]: Can we go with `dependencyChain.length > 0` ?

File: src/main/treeDataProviders/utils/yarnImpactGraph.ts [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+     * @param rawDependencyChain - The raw dependency chain string.
+     * @returns The extracted dependency chain or undefined if not found.
+     */
+    private extractChain(rawDependencyChain: string): string | undefined {
+        if (rawDependencyChain.toLocaleLowerCase().includes('specified in')) {
+            return this.dependencyName;
+        }
+        // Extract the path from the dependency chain using quotes
+        const startIndex: number = rawDependencyChain.indexOf('"');
+        const endIndex: number = rawDependencyChain.lastIndexOf('"');
```
[the remark this answers is not in the document]
Comment (Or-Geva) [reply src/main/treeDataProviders/utils/yarnImpactGraph.ts:136]: Changed to 
```
        const endIndex: number = rawDependencyChain.indexOf('"', startIndex);
```

**Pull request merged**